packages/state-transition/test/unit/cachedBeaconState.test.ts

```diff
@@ -174,7 +174,7 @@ describe("CachedBeaconState", () => {
         // confirm loadCachedBeaconState() result
         for (let i = 0; i < newCachedState.validators.length; i++) {
           expect(newCachedState.epochCtx.pubkey2index.get(newCachedState.validators.get(i).pubkey)).toBe(i);
-          expect(newCachedState.epochCtx.index2pubkey[i].toBytes()).toEqual(pubkeys[i]);
+          expect(Uint8Array.from(newCachedState.epochCtx.index2pubkey[i].toBytes())).toEqual(pubkeys[i]);
```

Comment:
Why is `Uint8Array.from()` needed here? `.toBytes()` should already return a `Uint8Array`

Reply:
I had an issue with that single test when we switched to vitest.  The vitest `toEqual` is a deeper equal than mocha/chai and Buffer != Uint8Array even though it inherits from it.

`toBytes` does not actually return a Uint8Array it returns a Buffer.  
type -
https://github.com/ChainSafe/blst-ts/blob/ac99092f57036abd03b9a1f8852c21c8a4db8c33/lib/index.d.ts#L111

implementation -
https://github.com/ChainSafe/blst-ts/blob/ac99092f57036abd03b9a1f8852c21c8a4db8c33/src/addon.h#L30

In order to return a Uint8Array from native one would need allocate the length of the backing store, and then manually copy the bytes.  

node-addon-api `TypedArrayOf` constructor definition -
https://github.com/nodejs/node-addon-api/blob/4e95635080fba645dda9d6d3e0611c40b163f5cf/napi.h#L1283

`Napi::Buffer::New()` allows passing the data pointer and length and then does the copy step internally.  

node-addon-api `Buffer` constructor definition -
https://github.com/nodejs/node-addon-api/blob/4e95635080fba645dda9d6d3e0611c40b163f5cf/napi.h#L1496

So I went with them instead.  Also allowed me to more easily implement `toHex` on each of the classes because on the `Buffer.toString` allows the encoding to be passed.  `Uint8Array.toString` always return `utf-8` I believe [uses `Array.prototype.toString`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/toString).

Reply:
fyi its because `fromHexString` in ssz returns a Uint8Array in the test setup

https://github.com/ChainSafe/lodestar/blob/f16a6f1c3a0c6fdbcaef8ff3061bd7dfb26f5337/packages/state-transition/test/utils/interop.ts#L16

![Uploading Screenshot 2024-04-17 at 2.41.18 PM.png…]()

Reply:
I did a bit more digging through commit history and using Buffer was a leftover from when I was having to copy bytes in one implementation....  I put up a PR in blst-ts that converts the serialized forms to `Uint8Array` for review.  Not sure if its a need to have at the moment and can get released with some other small changes (dual modules types, exporting the G1 and G2 zero values) that will get incorporated in the next publishing.

https://github.com/ChainSafe/blst-ts/pull/134.

This cast will get removed as well

Reply:
> toBytes does not actually return a Uint8Array it returns a Buffer.

I think we should avoid Buffer were possible, and although it's subclass of Uint8Array it overrides some methods and changes behavior, e.g. `Buffer.prototype.slice` does not copy memory but Uint8Array slice does. Also see https://github.com/nodejs/node/issues/41588#issuecomment-1016269584.

I think the node ecosystem in general wants to move avoid Buffer where possible.

Reply:
> Allowed me to more easily implement toHex on each of the classes because on the Buffer.toString allows the encoding to be passed. Uint8Array.toString always return utf-8 I believe [uses Array.prototype.toString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/toString).

There is a proposal for this https://github.com/tc39/proposal-arraybuffer-base64, but it will likely take a few more years until that's implemented and included in a node release.